zkevm-circuits/src/tx_circuit/sign_verify.rs

```diff
@@ -400,141 +348,213 @@ impl<F: Field, const MAX_VERIF: usize> SignVerifyChip<F, MAX_VERIF> {
+        // with RLC encoding corresponds to msg_hash_rlc
+        let msg_hash_rlc = {
+            let zero = chips.main_gate.assign_constant(ctx, F::zero())?;
+            let assigned_msg_hash_le = assigned_ecdsa
+                .msg_hash_le
+                .iter()
+                .map(|byte| chips.main_gate.select(ctx, &zero, byte, &is_address_zero))
+                .collect::<Result<Vec<_>, _>>()?;
+            let msg_hash_le = (!padding)
+                .then(|| sign_data.msg_hash.to_bytes())
+                .unwrap_or_default();
+            self.assign_rlc_le(
+                config,
+                ctx,
+                chips,
+                "msg_hash",
+                config.q_rlc_evm_word,
+                challenges.evm_word(),
+                assigned_msg_hash_le
+                    .iter()
+                    .zip(msg_hash_le)
+                    .map(|(assigned, byte)| {
+                        Term::assigned(assigned.cell(), Value::known(F::from(byte as u64)))
+                    }),
+            )?
+        };
 
-        let pk_be = pk_bytes_swap_endianness(&pk_le);
-        let mut keccak = Keccak::default();
-        keccak.update(&pk_be);
-        let pk_hash = keccak.digest();
-        let address = pub_key_hash_to_address(&pk_hash);
-
-        // Assign pk_hash
-        let pk_hash = if !padding { pk_hash } else { vec![0u8; 32] };
-        for (i, byte) in pk_hash.iter().enumerate() {
-            region.assign_advice(
-                || format!("pk_hash byte {}", i),
-                config.pk_hash[i],
-                offset,
-                || Value::known(F::from(*byte as u64)),
-            )?;
-        }
+        let pk_rlc = {
+            let assigned_pk_le = iter::empty()
+                .chain(&assigned_ecdsa.pk_y_le)
+                .chain(&assigned_ecdsa.pk_x_le);
+            let pk_le = iter::empty()
+                .chain(sign_data.pk.y.to_bytes())
+                .chain(sign_data.pk.x.to_bytes());
+            self.assign_rlc_le(
+                config,
+                ctx,
+                chips,
+                "pk_hash",
+                config.q_rlc_keccak_input,
+                challenges.keccak_input(),
+                assigned_pk_le.zip(pk_le).map(|(assigned, byte)| {
+                    Term::assigned(assigned.cell(), Value::known(F::from(byte as u64)))
+                }),
+            )?
+        };
 
-        let address = if !padding { address } else { F::zero() };
-        // Assign address and address_is_zero_chip
-        let address_assigned = region.assign_advice(
-            || "address",
-            config.address,
-            offset,
-            || Value::known(address),
+        let pk_hash_rlc = self.assign_rlc_le(
+            config,
+            ctx,
+            chips,
+            "pk_hash_rlc",
+            config.q_rlc_evm_word,
+            challenges.evm_word(),
+            iter::empty().chain(pk_hash_lo.into_iter().rev()).chain(
```

Comment:
Just to be sure: here we have all the terms in `pk_hash_lo` are `Term::assigned` and all the elements in `pk_hash_hi` `Term::unassigned` right?

Reply:
Yes, so `pk_hash_hi` are only witnessed for building the random linear combination of keccak input, but `pk_hash_lo` should be copied from `address` we have assigned ahead in order to get the value `is_address_zero`.